scripts/cat/cats.py

```diff
@@ -1559,39 +1568,220 @@ def thoughts(self):
         i = 0
         # for cats inside the clan
         if where_kitty == "inside":
-            while (
-                other_cat == self.ID
-                and len(all_cats) > 1
-                or (all_cats.get(other_cat).dead and dead_chance != 1)
-                or (other_cat not in self.relationships)
-            ):
-                other_cat = choice(list(all_cats.keys()))
-                i += 1
-                if i > 100:
-                    other_cat = None
-                    break
-        # for dead cats
+            # Roll relation and kin thought chances
+            if thought_type == 'relationship':
+                cats_with_relationship = [
+                    Cat.all_cats[rel_cat] for rel_cat in all_cats
+                    if rel_cat in self.relationships
+                    and Cat.all_cats[rel_cat].ID != self.ID
+                    and not (all_cats.get(Cat.all_cats[rel_cat].ID).dead and dead_chance != 1)
+                ]
+                if cats_with_relationship:
+                    chosen_cat = choice(cats_with_relationship)
+                    other_cat = chosen_cat.ID
+                    # print("relationship thought: from: ", self.name, " to: ", chosen_cat.name)
+                else:
+                    # if no cats available, roll again for a chance at a kin thought at a slightly higher chance
+                    # since we rolled a special thought already and couldn't get one
+                    thought_choices_reroll = ['kin', 'normal']
+                    weights_reroll = [
+                        (1.25 * game.config["relationship"]["kin_thought_chance"]),
+                        100 - game.config["relationship"]["kin_thought_chance"]
+                    ]
+                    thought_type = choices(thought_choices_reroll, weights=weights_reroll)[0]
+                
+            # Roll kin-based thought
+            if thought_type == 'kin':
+                kin_group_choice = randint(0, 100)
+                # If dead_chance == 1, include dead cats, otherwise exclude them
+                include_dead = (dead_chance == 1)
+                close_kin, kin, distant_kin = get_kin_groups(self, include_dead=include_dead)
+                
+                # 50% chance of choosing close kin
+                if kin_group_choice >= 50:
```

Comment:
The chances of each type of kin group can and should also be added in the settings.